cpp/include/raft/neighbors/brute_force_serialize.cuh

```diff
@@ -0,0 +1,209 @@
+ *
+ * raft::resources handle;
+ *
+ * // create an output stream
+ * std::ostream os(std::cout.rdbuf());
+ * // create an index with `auto index = brute_force::build(...);`
+ * raft::neighbors::brute_force::serialize(handle, os, index);
+ * @endcode
+ *
+ * @tparam T data element type
+ *
+ * @param[in] handle the raft handle
+ * @param[in] os output stream
+ * @param[in] index brute force index
+ *
+ */
+template <typename T>
+void serialize(raft::resources const& handle, std::ostream& os, const index<T>& index)
+{
+  RAFT_LOG_DEBUG(
+    "Saving brute force index, size %zu, dim %u", static_cast<size_t>(index.size()), index.dim());
+
+  auto dtype_string = raft::detail::numpy_serializer::get_numpy_dtype<T>().to_string();
+  dtype_string.resize(4);
+  os << dtype_string;
+
+  serialize_scalar(handle, os, serialization_version);
+  serialize_scalar(handle, os, index.size());
+  serialize_scalar(handle, os, index.dim());
+  serialize_scalar(handle, os, index.metric());
+  serialize_scalar(handle, os, index.metric_arg());
+  serialize_mdspan(handle, os, index.dataset());
+  auto has_norms = index.has_norms();
+  serialize_scalar(handle, os, has_norms);
+  if (has_norms) { serialize_mdspan(handle, os, index.norms()); }
+  resource::sync_stream(handle);
+}
+
+/**
+ * Save the index to file.
+ *
+ * Experimental, both the API and the serialization format are subject to change.
+ *
+ * @code{.cpp}
+ * #include <raft/core/resources.hpp>
+ * #include <raft/neighbors/brute_force_serialize.cuh>
+ *
+ * raft::resources handle;
+ *
+ * // create a string with a filepath
+ * std::string filename("/path/to/index");
+ * // create an index with `auto index = brute_force::build(...);`
+ * raft::neighbors::brute_force::serialize(handle, filename, index);
+ * @endcode
+ *
+ * @tparam T data element type
+ *
+ * @param[in] handle the raft handle
+ * @param[in] filename the file name for saving the index
+ * @param[in] index brute force index
+ *
+ */
+template <typename T>
+void serialize(raft::resources const& handle, const std::string& filename, const index<T>& index)
+{
+  auto os = std::ofstream{filename, std::ios::out | std::ios::binary};
+  RAFT_EXPECTS(os, "Cannot open file %s", filename.c_str());
+  serialize(handle, os, index);
+}
+
+/**
+ * Load index from input stream
+ *
+ * Experimental, both the API and the serialization format are subject to change.
+ *
+ * @code{.cpp}
+ * #include <raft/core/resources.hpp>
+ * #include <raft/neighbors/brute_force_serialize.cuh>
+ *
+ * raft::resources handle;
+ *
+ * // create an input stream
+ * std::istream is(std::cin.rdbuf());
+ * using T    = float; // data element type
+ * auto index = raft::neighbors::brute_force::deserialize<T>(handle, is);
+ * @endcode
+ *
+ * @tparam T data element type
+ *
+ * @param[in] handle the raft handle
+ * @param[in] is input stream
+ *
+ * @return raft::neighbors::brute_force::index<T>
+ */
+template <typename T>
+auto deserialize(raft::resources const& handle, std::istream& is)
```

Comment:
One thing I'd like to do here is provide an option to serialize out the entire dataset, similar to what we've done in CAGRA. This way, a user with a fairly sizable dataset doesn't need to write it out to the serialized file if they've already got it stored in a different format, but they do still benefit from serializing out the norms. THis should be fairly easy to add, but it's a pretty powerful feature.

Reply:
Love it! Will implement this right away